File: Course_Api/LAMS.Logic/Services/Teacher/TeacherService.cs
using AutoMapper;
using LAMS.DataAccess.Common.Models.Teacher;
using LAMS.DataAccess.Common.Repositories.Teacher;
using LAMS.Logic.Common.Models.Teacher;
using LAMS.Logic.Common.Services.Teacher;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LAMS.Logic.Services.Teacher
{
    public class TeacherService : ITeacherService
    {
        private IMapper _mapper;
        private ITeacherRepository _repo;

        public TeacherService(ITeacherRepository repo, IMapper mapper)
        {
            _repo = repo;
            _mapper = mapper;
        }

        public async Task<int> AddTeacherDirection(TeacherDirectionBLL direction)
        {
            try
            {
                var id = await _repo.AddTeacherDirection(_mapper.Map<TeacherDirectionDb>(direction)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public async Task<IEnumerable<TeacherDirectionBLL>> GetTeacherDirections(string id)
        {
            return await _repo.GetTeacherDirections(id)
                .ContinueWith(t => _mapper.Map<IEnumerable<TeacherDirectionBLL>>(t.Result));
        }

        public async Task<int> AddCourse(CourseBLL course)
        {
            try
            {
                var id = await _repo.AddCourse(_mapper.Map<CourseDb>(course)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public async Task<IEnumerable<CourseBLL>> GetCourses(int id)
        {
            return await _repo.GetCourses(id)
                .ContinueWith(t => _mapper.Map<IEnumerable<CourseBLL>>(t.Result));
        }
        public async Task<IEnumerable<CourseBLL>> GetActiveCourses()
        {
            return await _repo.GetActiveCourses()
                .ContinueWith(t => _mapper.Map<IEnumerable<CourseBLL>>(t.Result));
        }
        public async Task<CourseBLL> GetCourseInfo(int id)
        {
            return await _repo.GetCourseInfo(id)
                .ContinueWith(t => _mapper.Map<CourseBLL>(t.Result));
        }

        public async Task<int> FinishCreateCourse(int id)
        {
            return await _repo.FinishCreateCourse(_mapper.Map<int>(id)).ContinueWith(t => t.Result);
        }

        public async Task<int> StartCourse(int id)
        {
            return await _repo.StartCourse(_mapper.Map<int>(id)).ContinueWith(t => t.Result);
        }

        public async Task<int> EndCourse(int id)
        {
            return await _repo.EndCourse(_mapper.Map<int>(id)).ContinueWith(t => t.Result);
        }

        public async Task<int> EditCourse(CourseBLL course)
        {
            return await _repo.EditCourse(_mapper.Map<CourseDb>(course)).ContinueWith(t => t.Result);
        }
        public async Task<CourseBLL> DelCourse(int id)
        {
            return await _repo.DelCourse(id)
                .ContinueWith(t => _mapper.Map<CourseBLL>(t.Result));
        }
        public async Task<int> AddProgram(CourseProgramBLL program)
        {
            try
            {
                var id = await _repo.AddProgram(_mapper.Map<CourseProgramDb>(program)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public async Task<IEnumerable<CourseProgramBLL>> GetPrograms(int id)
        {
            return await _repo.GetPrograms(id)
                .ContinueWith(t => _mapper.Map<IEnumerable<CourseProgramBLL>>(t.Result));
        }
        public async Task<CourseProgramBLL> DelProgram(int id)
        {
            return await _repo.DelProgram(id)
                .ContinueWith(t => _mapper.Map<CourseProgramBLL>(t.Result));
        }
        public async Task<int> AddHomework(HomeworkBLL homework)
        {
            try
            {
                var id = await _repo.AddHomework(_mapper.Map<HomeworkDb>(homework)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public async Task<IEnumerable<HomeworkBLL>> GetHomeworks(int id)
        {
            return await _repo.GetHomeworks(id)
                .ContinueWith(t => _mapper.Map<IEnumerable<HomeworkBLL>>(t.Result));
        }
        public async Task<HomeworkBLL> DelHomework(int id)
        {
            return await _repo.DelHomework(id)
                .ContinueWith(t => _mapper.Map<HomeworkBLL>(t.Result));
        }

        public async Task<int> AddMaterial(MaterialBLL material)
        {
            try
            {
                var id = await _repo.AddMaterial(_mapper.Map<MaterialDb>(material)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public async Task<IEnumerable<MaterialBLL>> GetMaterials(int id)
        {
            return await _repo.GetMaterials(id)
                .ContinueWith(t => _mapper.Map<IEnumerable<MaterialBLL>>(t.Result));
        }
        public async Task<MaterialBLL> DelMaterial(int id)
        {
            return await _repo.DelMaterial(id)
                .ContinueWith(t => _mapper.Map<MaterialBLL>(t.Result));
        }
        public async Task<int> OrderCourse(LearnerCourseBLL course)
        {
            
                if (!await _repo.IsOrderCourse(course.IdCourse, course.IdUser))
                {
                    return 0;
                }
                var id = await _repo.OrderCourse(_mapper.Map<LearnerCourseDb>(course)).ContinueWith(t => t.Result);

                return id;
          
        }
        public async Task<IEnumerable<LearnerCourseBLL>> GetLearnerCourses(string id)
        {
            return await _repo.GetLearnerCourses(id)
                .ContinueWith(t => _mapper.Map<IEnumerable<LearnerCourseBLL>>(t.Result));
        }
        #region IDisposable Support
        private bool disposedValue = false; // Для определения избыточных вызовов

        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    // TODO: освободить управляемое состояние (управляемые объекты).
                }

                // TODO: освободить неуправляемые ресурсы (неуправляемые объекты) и переопределить ниже метод завершения.
                // TODO: задать большим полям значение NULL.

                disposedValue = true;
            }
        }

        // TODO: переопределить метод завершения, только если Dispose(bool disposing) выше включает код для освобождения неуправляемых ресурсов.
        // ~DocumentService() {
        //   // Не изменяйте этот код. Разместите код очистки выше, в методе Dispose(bool disposing).
        //   Dispose(false);
        // }

        // Этот код добавлен для правильной реализации шаблона высвобождаемого класса.
        void IDisposable.Dispose()
        {
            // Не изменяйте этот код. Разместите код очистки выше, в методе Dispose(bool disposing).
            Dispose(true);
            // TODO: раскомментировать следующую строку, если метод завершения переопределен выше.
            // GC.SuppressFinalize(this);
        }

        #endregion
    }
}

File: Couse_UI/src/app/shared/teacher.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from 'src/environments/environment';
import { map } from 'rxjs/operators';
import { User, UserReg } from './user.model';
import { Observable, Subject } from 'rxjs';
import { SaveResult } from './save-result.model';
import { TeacherDirection } from '../teacher-home/models/teacherdirection.model';
import { Course } from '../teacher-course/models/course.model';
import { Program } from '../teacher-programs/models/program.model';
import { Homework } from '../teacher-homework/models/homework.model';
import { Material } from '../teacher-material/models/material.model';
import { LearnerCourse } from '../user-course/models/learnercourse.model';

@Injectable({
  providedIn: 'root'
})
export class TeacherService {
  [x: string]: any;
  private URL_TEACHER = `${environment.apiHostUrl}/api/teacher`;

  constructor(
    private http: HttpClient
  ) { }

  AddTeacherDirection(newTeacherDirection: TeacherDirection): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_TEACHER + '/addteacherdirection';

    this.http.post(urlDoc, newTeacherDirection).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }

  getTeacherDirections(id): Observable<TeacherDirection[]> {
    const url = `${environment.apiHostUrl}/api/teacher/getteacherdirections?id=${id}`;
    return this.http
      .get<TeacherDirection[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new TeacherDirection(item.Id, item.IdUser, item.FIO, item.Experience, item.Workplace));
        })
      );
  }

  AddCourse(newCourse: Course): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_TEACHER + '/addcourse';

    this.http.post(urlDoc, newCourse).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  OrderCourse(newLearnerCourse: LearnerCourse): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_TEACHER + '/ordercourse';

    this.http.post(urlDoc, newLearnerCourse).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  getCourses(id): Observable<Course[]> {
    const url = `${environment.apiHostUrl}/api/teacher/getcourses?id=${id}`;
    return this.http
      .get<Course[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Course(item.Id, item.IdTeacherDirection, item.TeacherDirection, item.IdDirection,
            item.Direction, item.Name, item.Description, item.Skills, item.Status, item.DateStart));
        })
      );
  }
  getActiveCourses(): Observable<Course[]> {
    const url = `${environment.apiHostUrl}/api/teacher/getactivecourses`;
    return this.http
      .get<Course[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Course(item.Id, item.IdTeacherDirection, item.TeacherDirection, item.IdDirection,
            item.Direction, item.Name, item.Description, item.Skills, item.Status, item.DateStart));
        })
      );
  }
  getLearnerCourses(id: string): Observable<LearnerCourse[]> {
    const url = `${environment.apiHostUrl}/api/teacher/getlearnercourses?id=${id}`;
    return this.http
      .get<LearnerCourse[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new LearnerCourse(item.Id, item.IdCourse, item.IdUser, item.Name));
        })
      );
  }
  getCourseInfo(id): Observable<Course> {
    const url = `${environment.apiHostUrl}/api/teacher/getcourseinfo?id=${id}`;
    return this.http
      .get<Course>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result;
        })
      );
  }
  DeleteCourse(id: number): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const url = `${environment.apiHostUrl}/api/teacher/delcourse?id=${id}`;

    this.http.get(url).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  FinishCreateCourse(id: string): Observable<SaveResult> {
    const url = `${environment.apiHostUrl}/api/teacher/finishcreatecourse?id=${id}`;
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    this.http.get(url).subscribe(
      res => {
        result.next();
      },
      err => {
        result.next();
      }
    );
    return result.asObservable();
  }
  StartCourse(id: string): Observable<SaveResult> {
    const url = `${environment.apiHostUrl}/api/teacher/startcourse?id=${id}`;
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    this.http.get(url).subscribe(
      res => {
        result.next();
      },
      err => {
        result.next();
      }
    );
    return result.asObservable();
  }
  EditCourse(course: Course): Observable<number> {
     console.log(course);
    const result: Subject<number> = new Subject<number>();
    this.http.put(this.URL_TEACHER + '/editcourse', course).subscribe(
      res => {
          result.next();
      },
      err => {
        result.next();
      }
    );
    return result.asObservable();
  }
  AddProgram(newProgram: Program): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_TEACHER + '/addprogram';

    this.http.post(urlDoc, newProgram).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  getPrograms(id): Observable<Program[]> {
    const url = `${environment.apiHostUrl}/api/teacher/getprograms?id=${id}`;
    return this.http
      .get<Program[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Program(item.Id, item.IdCourse, item.Theme));
        })
      );
  }
  DeleteProgram(id: number): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const url = `${environment.apiHostUrl}/api/teacher/delprogram?id=${id}`;

    this.http.get(url).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  AddHomework(newHomework: Homework): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_TEACHER + '/addhomework';

    this.http.post(urlDoc, newHomework).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  getHomeworks(id): Observable<Homework[]> {
    const url = `${environment.apiHostUrl}/api/teacher/gethomeworks?id=${id}`;
    return this.http
      .get<Homework[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Homework(item.Id, item.IdCourse, item.Link, item.Description));
        })
      );
  }
  DeleteHomework(id: number): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const url = `${environment.apiHostUrl}/api/teacher/delhomework?id=${id}`;

    this.http.get(url).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  AddMaterial(newMaterial: Material): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_TEACHER + '/addmaterial';

    this.http.post(urlDoc, newMaterial).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  getMaterials(id): Observable<Material[]> {
    const url = `${environment.apiHostUrl}/api/teacher/getmaterials?id=${id}`;
    return this.http
      .get<Material[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Material(item.Id, item.IdCourse, item.Link, item.Description));
        })
      );
  }
  DeleteMaterial(id: number): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const url = `${environment.apiHostUrl}/api/teacher/delmaterial?id=${id}`;

    this.http.get(url).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  private getOptions(query: string, status: string) {
    return {
      params: new HttpParams()
        .set('UserId', query)
        .set('Status', status)
    };
  }
}
